### ISSUE TITLE
Bug: not able to find config files

### ISSUE THREAD

**Issue by artola** [typescript-eslint/typescript-eslint#6883]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

typescript-estree

### Repro Code

```typescript
The logic in `getProjectConfigFiles` compares the length of `parseSettings.filePath` (dirname) and `parseSettings.tsconfigRootDir` to stop looking for the config.

But when `tsconfigRootDir` is not set in the `.eslintrc.js` config, they can be totally different paths, for example, depending where VSCode was installed (value used for `tsconfigRootDir`) for tooling.

Then, it is correct to stop looking for config "only" if the `parseSettings.filePath` starts with `tsconfigRootDir`, otherwise just keep looking up to the root.

https://github.com/typescript-eslint/typescript-eslint/blob/a4b633b9ab106e5889c2372c018b1b937ff0afbb/packages/typescript-estree/src/parseSettings/getProjectConfigFiles.ts#L53
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/<rule-name>": ["error", ...<options>],
  },
  parserOptions: {
    project: true
  }
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

The config file should be found.

### Actual Result

The config file is not found because it stops prematurely.

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `5.58.0` |
| `@typescript-eslint/parser`            | `5.58.0` |
| `@typescript-eslint/scope-manager`     | `5.58.0` |
| `@typescript-eslint/typescript-estree` | `5.58.0` |
| `@typescript-eslint/type-utils`        | `5.58.0` |
| `@typescript-eslint/utils`             | `5.58.0` |
| `TypeScript`                           | `5.0.4` |
| `ESLint`                               | `8.38.0` |
| `node`                                 | `18.14.1` |

**Comment by bradzacher** [comment]
`tsconfigRootDir === process.cwd()` if you do not set a value for it.
Your IDE extension should be setting the CWD to the workspace root when it starts the lint run.

Is that not happening? It's really hard to tell what's happening here without some more information.
Could you please provide a repro case to help us understand and reproduce the problem?

**Comment by bradzacher** [comment]
I'm not able to repro that behaviour with VSCode.
I just added a log to the start of the parser (`console.log('### CWD: ', process.cwd());`) and then restarted the extension, and it logs out root path of the project - not the install path of VSCode.
```ts
[Error - 5:33:38 PM] Linting file file:///e%3A/github/typescript-eslint/packages/eslint-plugin/src/index.ts took 23369ms. Please check the ESLint rules for performance issues.
### CWD:  E:\github\typescript-eslint
```

---

To be clear - not saying this isn't something that we should consider fixing in some way - I'm just trying to correctly narrow down a repro so that we can fully understand the real-world usecase and impact of the problem.

**Comment by artola** [comment]
@bradzacher Here a repo for this issue: https://github.com/artola/typescript-estree-issue

Once cloned and installed using `yarn` (other package manager should be no problem either), I did modify this file `node_modules/@typescript-eslint/typescript-estree/dist/parseSettings/getProjectConfigFiles.js`

Exactly at line 61 I wrote this: `throw new Error(`===> ${JSON.stringify(parseSettings, null, 2)}`);`.

I did reload VSCode and visit `src/main.ts` after that run the command `Typescript: Open TS Server log`.

On Mac it shows something like:

```text
Info 142  [11:10:52.877] [typescript-eslint-language-service] Error: ===> {
  "code": "debugger;\nconst x = {    y: 123   } ",
  "comment": true,
  "comments": [],
  "createDefaultProgram": false,
  "debugLevel": {},
  "errorOnTypeScriptSyntacticAndSemanticIssues": false,
  "errorOnUnknownASTType": false,
  "EXPERIMENTAL_useSourceOfProjectReferenceRedirect": false,
  "extraFileExtensions": [],
  "filePath": "/Users/martin/Vault/spaces/OSS/typescript-eslint-language-service-issue/src/main.ts",
  "jsx": false,
  "loc": true,
  "moduleResolver": "",
  "preserveNodeMaps": true,
  "programs": null,
  "projects": [],
  "range": true,
  "singleRun": false,
  "tokens": [],
  "tsconfigMatchCache": {},
  "tsconfigRootDir": "/"
}
```

As you can see `"tsconfigRootDir": "/"` matches the start for my file ` "filePath": "/Users/martin/Vault/spaces/OSS/typescript-eslint-language-service-issue/src/main.ts",`.

But doing the same on Windows returns:

```text
Info 73   [11:22:31.937] [typescript-eslint-language-service] ===> {
  "code": "debugger;\r\nconst x = {    y: 123   }",
  "comment": true,
  "comments": [],
  "createDefaultProgram": false,
  "debugLevel": {},
  "errorOnTypeScriptSyntacticAndSemanticIssues": false,
  "errorOnUnknownASTType": false,
  "EXPERIMENTAL_useSourceOfProjectReferenceRedirect": false,
  "extraFileExtensions": [],
  "filePath": "c:/Users/Martin/Vault/spaces/OSS/typescript-eslint-language-service-issue/src/main.ts",
  "jsx": false,
  "loc": true,
  "moduleResolver": "",
  "preserveNodeMaps": true,
  "programs": null,
  "projects": [],
  "range": true,
  "singleRun": false,
  "tokens": [],
  "tsconfigMatchCache": {},
  "tsconfigRootDir": "C:\\Users\\Martin\\AppData\\Local\\Programs\\Microsoft VS Code"
}
```

Here not only they do not match, also they contain different separator.

It is not clear to me why/where the mismatching is introduced. Anyway, as you can see relying in the `length` of the paths might be not enough.

**Comment by bradzacher** [comment]
Ah - that's the missing piece that was really important to reproducing this:

```
    "plugins": [{ "name": "typescript-eslint-language-service" }]
```

You're not using our tooling via the standard vscode-eslint extension which is built to run eslint and correctly sets the CWD to the project folder.

You're using our tooling via a 3rd party TS language server plugin. I assume that the TS integration with vscode doesn't set the CWD to the project folder when it spawns the language server then?

Thankfully at least this issue is going to be a rarer case because that integration is a fraction of a percent of our overall user base.

**Comment by artola** [comment]
@bradzacher Right, I am using `typescript-eslint-language-service`, I find a language service really awesome for performance and integration of third party tooling.

**Comment by JoshuaKGoldberg** [comment]
Ah, we're going to have to get more knowledgeable on that plugin. I've been meaning to look more deeply at https://github.com/Quramy/typescript-eslint-language-service/issues/497#issuecomment-1500913573 too. Thanks for filing!

**Comment by Josh-Cena** [comment]
How may we fix this in our code? Should the language server be setting the CWD?

**Comment by bradzacher** [comment]
> Using tsconfigRootDir: process.cwd() works a bit better than tsconfigRootDir: __dirname when the .eslintrc.js is extending another config file that might be inside node_modules as a NPM package.

I'm not sure what you mean by this. As it is written - it is non-sensical because:
- ESLint fully resolves the config completely disconnected from the existence of `node_modules`
   - it resolves `prettier` -> `node_modules/eslint-config-prettier`, loads it, merges the configs together and thus forgets that the config ever came from `node_modules`.
- `tsconfigRootDir` ***only*** applies to our parser if and only if `parserOptions.project != null`.
- The default value for `tsconfigRootDir` is `process.cwd()` - so explicitly setting it to `process.cwd()` is the same as not setting it at all.

The only solution currently is to use `tsconfigRootDir: __dirname` so that we know the exact path to your project root.
You should be doing this regardless because:
- it's a performance optimisation when `project: true` - we know exactly when to stop traversing
- it fixes weirdness with `project: string | string[]` because it means relative paths are resolved absolutely, regardless of the CWD.

---

I'm not exactly sure - TBH - this is really into "undefined" territory.

If you set `project: true`, but DON'T set `tsconfigRootDir` and `!filePath.includes(cwd)`, what should we do?

I can only see two options:
1) Crash out and say "hey you're being weird - please set `tsconfigRootDir` to `__dirname`.".
    - This is, in essence, what we _implicitly_ do right now.
2) Ignore the CWD and treat it as `/` (i.e. traverse all the way to the disk root)

(2) feels bad, IMO - it feels like it's setting up the user for slow lint runs by default.
I'd personally prefer just crashing out so that we can enforce a better standard.

**Comment by bradzacher** [comment]
@artola to clarify - you should *never* set `tsconfigRootDir` from within a shared config and only ever from within a `.eslintrc.js` file. EG:
```ts
module.exports = {
  extends: ['local-shared'],
  parserOptions: {
    tsconfigRootDir: __dirname,
  },
};
```

This is because, as mentioned before, the intention behind the option is that it gives us a physical root to orient ourselves relative to your project root. Setting it within a shared config that lives god knows where is entirely useless, as you've mentioned.

**Comment by artola** [comment]
@bradzacher As mentioned above, my actual config looks like:

```javascript
module.exports = {
  extends: ['local-shared'],
  parserOptions: {
    project: true,
    tsconfigRootDir: __dirname,
  },
};
```

And it works well on Mac and Windows. I did open this issue, because `tsconfigRootDir` is not set, it "strangely" sometimes works or fails depending where it is installed VSCode or if it is running on Mac or Windows.
